### PR TITLE
restrict fonts on certain linux distros

### PR DESCRIPTION
**Pull request by arthuredelstein**
Resolves https://github.com/brave/brave-browser/issues/34169

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: third_party/blink/renderer/brave_font_whitelist.cc [hunk 1]
```diff
@@ -9,13 +9,22 @@
...
+#include "base/linux_util.h"
 
 namespace brave {
 
 namespace {
 
 bool g_simulate_empty_font_whitelist_for_testing = false;
 
+#if BUILDFLAG(IS_LINUX)
+const auto g_distro = base::GetLinuxDistro();
```
Comment (goodov) [inline third_party/blink/renderer/brave_font_whitelist.cc:21]: this is not great. You don't need this global variable at the very start of the process. Consider a more startup-friendly approach:
```
const std::string& GetLinuxDistro() {
  static const base::NoDestructor<std::string> linux_distro(base::GetLinuxDistro());
  return *linux_distro;
}
```

File: third_party/blink/renderer/brave_font_whitelist.cc [hunk 2]
```diff
@@ -9,13 +9,22 @@
...
+#endif
+
+constexpr char kUbuntu2204Prefix[] = "Ubuntu 22.04";
+constexpr char kUbuntu2004Prefix[] = "Ubuntu 20.04";
+constexpr char kFedora32Prefix[] = "Fedora";
```
Comment (goodov) [inline third_party/blink/renderer/brave_font_whitelist.cc:26]: you can replace this with `enum class`, do all comparison in the function and return only enum value.

File: third_party/blink/renderer/brave_font_whitelist.cc [hunk 3]
```diff
@@ -955,15 +2046,12 @@ base::span<const std::string_view> GetAdditionalFontWhitelistByLocale(
...
   return kEmptyFontSet;
 }
 
 size_t GetFontWhitelistSizeForTesting() {
-  return kFontWhitelist.size();
-}
-
-bool CanRestrictFontsForTesting() {
-  return kCanRestrictFonts;
+  return GetMainFontWhitelist().size();
```
Comment (goodov) [inline third_party/blink/renderer/brave_font_whitelist.cc:2054]: `!empty()`

Comment (arthuredelstein) [reply third_party/blink/renderer/brave_font_whitelist.cc:2054]: For the unit tests, we're actually interested in the size, not just whether the list is empty. See `third_party/blink/renderer/brave_font_whitelist_unittest.cc`

File: third_party/blink/renderer/brave_font_whitelist.cc (hunk 2 above)
Comment (arthuredelstein) [reply third_party/blink/renderer/brave_font_whitelist.cc:26]: Done

File: third_party/blink/renderer/brave_font_whitelist.cc (hunk 1 above)
Comment (arthuredelstein) [reply third_party/blink/renderer/brave_font_whitelist.cc:21]: For some reason when I don't call this function early in the process, it always returns "Unknown" for the Linux distro. Whereas if I call it at process start, then I get the correct string "Ubuntu 22.04.3 LTS". So far I'm not sure why this happens, but I'm wondering if you think an early call could be acceptable under the circumstances.

File: third_party/blink/renderer/brave_font_whitelist.cc [hunk 4]
```diff
@@ -18,13 +19,35 @@ namespace {
...
 constexpr char kFedora32Prefix[] = "Fedora";
 
+enum class LinuxDistro {
+  kUnknown,
+  kUbuntu2204,
+  kUbuntu2004,
+  kFedora32,
+};
+
+LinuxDistro GetLinuxDistro() {
```
Comment (goodov) [inline third_party/blink/renderer/brave_font_whitelist.cc:34]:
```
LinuxDistro GetLinuxDistro() {
  static const LinuxDistro kLinuxDistro = []() {
    const auto distro = base::GetLinuxDistro();
    if (distro.starts_with(kUbuntu2204Prefix)) {
      return LinuxDistro::kUbuntu2204;
    } else if (distro.starts_with(kUbuntu2004Prefix)) {
      return LinuxDistro::kUbuntu2004;
    } else if (distro.starts_with(kFedora32Prefix)) {
      return LinuxDistro::kFedora32;
    } else {
      return LinuxDistro::kUnknown;
    }
  }();
  return kLinuxDistro;
}
```

and remove `const auto g_distro = GetLinuxDistro();`.

File: third_party/blink/renderer/brave_font_whitelist.cc [hunk 5]
```diff
@@ -1949,13 +1972,15 @@ constexpr auto kFontWhitelistML = kEmptyFontSet;
 
 base::span<const std::string_view> GetMainFontWhitelist() {
 #if BUILDFLAG(IS_LINUX)
-  if (g_distro.starts_with(kUbuntu2204Prefix)) {
+  const auto distro = GetLinuxDistro();
+  printf("distro: %d\n", static_cast<int>(distro));
```
Comment (goodov) [inline third_party/blink/renderer/brave_font_whitelist.cc:1976]: leftovers?

File: third_party/blink/renderer/brave_font_whitelist.cc [hunk 6]
```diff
@@ -7,6 +7,7 @@
 
 #include <string_view>
 #include <vector>
+#include "stdio.h"
```
Comment (goodov) [inline third_party/blink/renderer/brave_font_whitelist.cc:10]: leftovers?

**Review by goodov — APPROVED** [review]

File: third_party/blink/renderer/brave_font_whitelist.cc (hunk 6 above)
Comment (arthuredelstein) [reply third_party/blink/renderer/brave_font_whitelist.cc:10]: Removed

File: third_party/blink/renderer/brave_font_whitelist.cc (hunk 5 above)
Comment (arthuredelstein) [reply third_party/blink/renderer/brave_font_whitelist.cc:1976]: Removed

File: third_party/blink/renderer/brave_font_whitelist.cc (hunk 1 above)
Comment (arthuredelstein) [reply third_party/blink/renderer/brave_font_whitelist.cc]: Addressed in https://github.com/brave/brave-core/pull/20930#discussion_r1395218706

**Comment by arthuredelstein** [comment]
Thanks for the review!

**Pull request merged**